scrapy/commands/shell.py

```diff
@@ -47,7 +50,17 @@ def update_vars(self, vars):
     def run(self, args, opts):
         url = args[0] if args else None
         if url:
-            url = add_http_if_no_scheme(url)
+            parts = urlparse(url)
+            if not parts.scheme:
+                if "." not in parts.path.split("/", 1)[0]:
+                    url = any_to_uri(url)
+
+                for pattern in ["/", "./", "../"]:
+                    if url.startswith(pattern):
+                        url = any_to_uri(url)
+                        break
+                url = add_http_if_no_scheme(url)
```

Comment:
I think this code should be extracted to a helper function, it will make testing easier.